packages/flutter/lib/src/rendering/semantics.dart

```diff
@@ -11,6 +11,8 @@ import 'package:flutter/painting.dart';
 import 'package:vector_math/vector_math_64.dart';
 
 import 'node.dart';
+import 'object.dart';
+import 'viewport.dart';
```

Comment:
Sure, feel free to copy-paste the above into a README or object.dart or whatever. Not sure where it should go.

Sorry it wasn't written down before, it kind of grew organically in Adam and I's head.

FWIW, the other libraries have similar inner layers, e.g. widget has framework.dart at the base, with basic.dart on top, and everything else on top of that. In theory, anyway; that one has some cross-talk currently via debug.dart.

Fundamentally though this is just a manifestation of the more concrete best practice of avoiding reference cycles. It's a general truism that a good architecture has a dependency graph that's a DAG, with the units being as small as possible (ideally much smaller than 1000 lines of Dart, based on our experience so far). Cycles usually indicate a bug factory. Files larger than 1000 lines usually indicate a bug factory. (And yes, we have plenty of bug factories, unfortunately. We should try to minimise them.) This is one reason we stopped using `part`, it encourages designs that turn into bug factories.

I've considered trying to enforce our dependency graph being a DAG. We do enforce it at the library level now (e.g. widgets can't depend on material). I suspect that enforcing it at the library level today would involve a few weeks of disentangling. cc @a14n in case you're interested in creating a lint for this. :-)